From cce91bd5a8f908a40b032b0d13f15dde0da9e66b Mon Sep 17 00:00:00 2001
From: Benjamin Maxwell <benjamin.maxwell@arm.com>
Date: Wed, 15 Apr 2026 16:39:03 +0000
Subject: [PATCH] Add minimal SME ABI routines (__arm_sme_state)

Note: We cache "has_sme" to avoid checking id_aa64pfr1_el1 in user code.
---
 libc/machine/aarch64/CMakeLists.txt |  1 +
 libc/machine/aarch64/meson.build    |  1 +
 libc/machine/aarch64/sme-abi.S      | 30 +++++++++++++++++++++++++++++
 picocrt/machine/aarch64/crt0.c      | 20 +++++++++++++++++++
 4 files changed, 52 insertions(+)
 create mode 100644 libc/machine/aarch64/sme-abi.S

diff --git a/libc/machine/aarch64/CMakeLists.txt b/libc/machine/aarch64/CMakeLists.txt
index f0e48595d..0736785db 100644
--- a/libc/machine/aarch64/CMakeLists.txt
+++ b/libc/machine/aarch64/CMakeLists.txt
@@ -68,6 +68,7 @@ picolibc_sources_flags("-fno-builtin"
   strnlen.S
   strrchr-stub.c
   strrchr.S
+  sme-abi.S
   tls.c
   )
 
diff --git a/libc/machine/aarch64/meson.build b/libc/machine/aarch64/meson.build
index 8a11d7d15..e5ee1989a 100644
--- a/libc/machine/aarch64/meson.build
+++ b/libc/machine/aarch64/meson.build
@@ -70,6 +70,7 @@ srcs_machine = [
   'strnlen.S',
   'strrchr-stub.c',
   'strrchr.S',
+  'sme-abi.S',
   'tls.c'
 ]
 
diff --git a/libc/machine/aarch64/sme-abi.S b/libc/machine/aarch64/sme-abi.S
new file mode 100644
index 000000000..c5c2f9b02
--- /dev/null
+++ b/libc/machine/aarch64/sme-abi.S
@@ -0,0 +1,30 @@
+/*
+ * sme-abi.S - minimal SME ABI routines implementation
+ *
+ * Copyright (c) 2026, Arm Limited.
+ * SPDX-License-Identifier: MIT OR Apache-2.0 WITH LLVM-exception
+ */
+
+#include "asmdefs.h"
+
+// __arm_sme_state fills the result registers based on id_aa64pfr1_el1.
+ENTRY (__arm_sme_state)
+  .variant_pcs __arm_sme_state
+  mov x0, xzr
+  mov x1, xzr
+
+  /* Check if we have SME (global initialized in crt0.c) */
+  adrp x16, __arm64_has_sme
+  ldrb w16, [x16, :lo12:__arm64_has_sme]
+  cmp w16, #0
+  b.eq	L(no_sme)
+
+  orr x0, x0, #0xC000000000000000
+  /* Read SVCR */
+  mrs x16, s3_3_c4_c2_2
+  bfxil x0, x16, #0, #2
+  /* Read TPIDR2_EL0 */
+  mrs x1, s3_3_c13_c0_5
+L(no_sme):
+  ret
+END (__arm_sme_state)
diff --git a/picocrt/machine/aarch64/crt0.c b/picocrt/machine/aarch64/crt0.c
index a43092ee9..d16454f9d 100644
--- a/picocrt/machine/aarch64/crt0.c
+++ b/picocrt/machine/aarch64/crt0.c
@@ -61,6 +61,26 @@ _set_tls(void *tls)
     __asm__ volatile("msr tpidr_el0, %x0" : : "r"(tls - TP_OFFSET));
 }
 
+#define FEAT_SME_1      (1LL << 24)
+#define FEAT_SME_2      (1LL << 25)
+
+_Bool __arm64_check_has_sme_or_sme2_el1() {
+    /* Determine if SME or SME2 is available */
+    int result = 0;
+    __asm__ volatile ( "mrs %x[result], id_aa64pfr1_el1"
+        : [result]"=r"(result) :  : );
+    return (result & (FEAT_SME_1|FEAT_SME_2)) != 0;
+}
+
+_Bool __arm64_has_sme;
+
+static void __arm64_post_memory_setup() {
+    /* Initialize __arm64_has_sme once .bss and .data have been setup. */
+    __arm64_has_sme = __arm64_check_has_sme_or_sme2_el1();
+}
+
+#define POST_MEMORY_SETUP __arm64_post_memory_setup
+
 #include "../../crt0.h"
 
 /* Defined in crt0.S */
-- 
2.43.0

